### DISCUSSION TITLE
Custom Pydantic models are not recognized as body parameters

### DISCUSSION THREAD

**Discussion by drhagen** [fastapi/fastapi#8599] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Body
from fastapi.testclient import TestClient

app = FastAPI()

class NastyMetaClass(type):
    pass

class Foo(metaclass=NastyMetaClass):
    @classmethod
    def __get_validators__(cls):
        yield lambda value: True

@app.post("/implicit")
def foo(foo: Foo):  # This is supposed to work, but does not
    return "It worked"

@app.post("/explicit")
def foo_body(foo: Foo = Body(...)):  # The `= Body(...)` fixes it
    return "It worked"

client = TestClient(app)

response = client.post("/implicit", json={})
print(response.json())
# {'detail': [{'loc': ['query', 'foo'], 'msg': 'field required', 'type': 'value_error.missing'}]}

response = client.post("/explicit", json={})
print(response.json())
# It worked
```

### Description

[Pydantic says](https://pydantic-docs.helpmanual.io/usage/types/#custom-data-types) that you can create custom classes by simply defining the `__get_validators__` method. This is useful if you want to parse into a class with its own metaclass or for some other reason do not want to inherit from `BaseModel`.

Models defined that way work most places in FastAPI, but not everywhere. [For example](https://stackoverflow.com/questions/71177433/custom-pydantic-classes-as-fastapi-body-parameters#71177977), FastAPI does not detect such a class as a body parameter, but always thinks it is a query parameter. This can be worked around by putting `= Body(...)` on each use.

For the body parameter example, I traced it down to [this](https://github.com/tiangolo/fastapi/blob/5a864d826139c09c42d1c0f504c232bcf2ce8f54/fastapi/dependencies/utils.py#L219) line `lenient_issubclass(field.type_, BaseModel)` in `utils.py`. If this were replaced with `hasattr(field.type_, '__get_validators__')` then it would work as I expected.

Now, this could be something that FastAPI expects Pydantic to handle with a `__subclasscheck__` on `ModelMetaclass` or whatever, but I thought I would report it here in case FastAPI intended to support all Pydantic models.

### Operating System

Linux

### FastAPI Version

0.74.0

### Python Version

3.9.6

**Comment by mohsen-mahmoodi** [comment]
I think there is no mismatch between the documentation and the implementation of FastAPI, FastAPI relies on some common base classes for some of its behaviors e.g. auto doc generation, you can dig into the code to see the usage and conversion of request body to `BaseModel`. Also in the documentation, it is mentioned that you can use scaler values as Body by using the `= Body(...)`. 

Using `__get_validators__` only works well for model fields and if you look at the Pydantic example, you will see that the example is inheriting from `str`. Now in your example, if you inherit from scaler types like `str`, `int`, etc. the `foo` will be considered as a query parameter, if you inherit from collections like `dict`, `list` or the `BaseModel` FastAPI will consider it as Body as mentioned in documentation. 

Now considering that FastAPI relies on `BaseModel` features, relying on just `__get_validators__` will break some part of the implementation, for example, now you should define `__modify_schema__` and other built-in methods of `BaseModel` so that FastAPI works as expected, e.g. the auto doc generation feature, etc.

Still you can consider it as the default behavior of FastAPI, it considers scaler values as query params, and if you need to use a custom class as a request body either you can make your class none-scaler or just use the `= Body(...)`  solution. 

And FastAPI hasn't intended to support all Pydantic classes as request body by default. It intended to support Pydantic models as you said.

**Reply by armanschwarz** [reply]
> I think there is no mismatch between the documentation and the implementation of FastAPI

The documentation states: 

> If the parameter is declared to be of the type of a Pydantic model, it will be interpreted as a request body.

This is not how `fastapi` behaves.
